graph-builder/src/registry.rs

```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+fn trim_protocol(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+pub fn authenticate_client<'a>(
+    client: &'a mut dkregistry::v2::Client,
+    login_scope: std::string::String,
+) -> impl futures::future::Future<Item = &'a dkregistry::v2::Client, Error = dkregistry::errors::Error>
+{
+    client
+        .is_v2_supported()
+        .and_then(move |v2_supported| {
+            if !v2_supported {
+                Err("API v2 not supported".into())
+            } else {
+                Ok(client)
+            }
+        })
+        .and_then(|dclient| {
+            dclient.is_auth(None).and_then(move |is_auth| {
+                if is_auth {
+                    Err("no login performed, but already authenticated".into())
+                } else {
+                    Ok(dclient)
+                }
+            })
+        })
+        .and_then(move |dclient| {
+            dclient.login(&[&login_scope]).and_then(move |token| {
+                dclient
+                    .is_auth(Some(token.token()))
+                    .and_then(move |is_auth| {
+                        if !is_auth {
+                            Err("login failed".into())
+                        } else {
+                            Ok(dclient.set_token(Some(token.token())))
+                        }
+                    })
+            })
+        })
+}
+
 /// Fetches a vector of all release metadata from the given repository, hosted on the given
 /// registry.
-pub fn fetch_releases(registry: &str, repo: &str) -> Result<Vec<Release>, Error> {
-    let mut metadata = Vec::new();
-    for tag in fetch_tags(registry, repo)? {
-        metadata.push(Release {
-            source: format!(
-                "{}/{}:{}",
-                registry
-                    .trim_left_matches("https://")
-                    .trim_left_matches("http://"),
-                repo,
-                tag
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_path: &Option<PathBuf>,
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(|| -> Result<Box<dyn std::io::Read>, Error> {
+            match credentials_path {
+                Some(path) => match File::open(path) {
```

Comment:
I don't think we should re-read this file every time we want to fetch the releases. Can we move this into main?

Reply:
The read happens within the library so if we want to avoid re-read we have to move the client creation to main. I don't think it's very expensive as it's certainly buffered after the first read.

Reply:
I think this is okay for now, but I think we should eventually move it.